### PR TITLE
Properly escape IDs in `getSelector()` to handle weird IDs

### LINKED ISSUE #31099: Dynamic Tabs, where href starts with number, not working
Example: https://jsfiddle.net/wcq4L2fr/1/

When the nav-link href starts with a number, or is only a number (`<a class="nav-link" data-toggle="tab" href="#1">`), it will not expose the referenced contents (`<div class="tab-pane fade" id="1">`).

If the nav-link href is simply alpha, or begins with an alpha character, it works fine.

### LINKED ISSUE #34381: Scrollspy does not handle all valid HTML5 element ids
## Bug Report

Scrollspy stops working when any of the DOM elements it is "spying" on has an id that starts with a number, such as

```html
<div id="3rd">Text</div>
```

This should be considered a bug because HTML5 has no restrictions on what form an ID can take ([official spec](https://html.spec.whatwg.org/multipage/dom.html#global-attributes:the-id-attribute-2)).
> There are no other restrictions on what form an ID can take; in particular, IDs can consist of just digits, start with a digit, start with an underscore, consist of just punctuation, etc.

Historically, HTML4 did not allow for ids to start with numbers, but it would seem ridiculous to assume only HTML4 content today.

## Bug Demo

See this JSBin for the demo. https://jsbin.com/rehematazo/3/edit?html,output. By default, I've commented out the `<a>` elements that link to `<div>`s whose IDs start with a digit, so Scrollspy works. But once you uncomment out these `<a>` tags, then Scrollspy stops working entirely.

I have also copied the HTML code for the demo here:
<details>
<summary>Code</summary>

```html
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width">
  <title>JS Bin</title>
  <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-EVSTQN3/azprG1Anm3QDgpJLIm9Nao0Yz1ztcQTwFspd3yD65VohhpuuCOmLASjC" crossorigin="anonymous">
  <style>
    .tall { height: 300px; }
    
    .toc-nav { /* <nav> */
      position: sticky;
      top: 0;
      height: min-content;
      max-height: 100vh;
      overflow-y: auto;
    }
    
    .toc-nav a.active {
      color: #007bff;
    }
  </style>
</head>
<body data-bs-spy="scroll" data-bs-target="#nav">

  <div class="container"><div class="row">

  <main class="col-8">
  <h2 id="A">ID A</h2>
  <div class="tall">Chunk of text</div>
  <h2 id="B">ID B</h2>
  <div class="tall">Chunk of text</div>
  <h2 id="C">ID C</h2>
  <div class="tall">Chunk of text</div>

  <h2 id="1">ID 1</h2>
  <div class="tall">Chunk of text</div>
  <h2 id="2">ID 2</h2>
  <div class="tall">Chunk of text</div>
  <h2 id="3">ID 3</h2>
  <div class="tall">Chunk of text</div>
  </main>
  
  <nav class="col-4 toc-nav" id="nav">
    <nav class="navbar navbar-light">
      <span class="navbar-brand">Contents</span>
      <a class="list-group-item list-group-item-action" href="#A">Item A</a>
      <a class="list-group-item list-group-item-action" href="#B">Item B</a>
      <a class="list-group-item list-group-item-action" href="#C">Item C</a>

      <!-- Uncomment the following 3 lines to see the bug:
           Once you uncomment, you will notice that Scrollspy stops working.
        -->
      <!-- <a class="list-group-item list-group-item-action" href="#1">Item 1</a>
      <a class="list-group-item list-group-item-action" href="#2">Item 2</a>
      <a class="list-group-item list-group-item-action" href="#3">Item 3</a> -->
    </nav>
  </nav>

  </div></div>

  <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/js/bootstrap.min.js" integrity="sha384-cVKIPhGWiC2Al4u+LWgxfKTRIcfu0JTxR+EQDz/bgldoEyl4H0zUF0QKbrJ0EcQF" crossorigin="anonymous"></script>
</body>
</html>
```

</details>

## Why this happens, and possible fixes

This bug seems to happen because Bootstrap uses [`document.querySelector()`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector) to identify elements:

https://github.com/twbs/bootstrap/blob/696996dd4f5b7319413655b9134ca2f3768c2549/js/dist/scrollspy.js#L78

However, `document.querySelector()` only works for valid CSS selector strings, and CSS selector strings for IDs cannot start with a number (even though HTML IDs can). See this StackOverflow Q&A for more info: https://stackoverflow.com/q/20306204. This causes Bootstrap JS to encounter the following error (copied from Microsoft Edge console):

```
Uncaught DOMException: Failed to execute 'querySelector' on 'Document': '#3rd' is not a valid selector.
    at <anonymous>:1:10
```

The StackOverflow Q&A proposes several possible fixes. For example, `document.getElementById("3rd")` works fine, but only works with ID-based selectors. I am also not sure which fixes fall within Bootstrap's browser compatibility goals.

## System Info

OS: Windows 10 20H1
Browser: Microsoft Edge (Chromium) 91.0.864.59 and Firefox 89.0.2 (64-bit)
Bootstrap version 5.0.2

### LINKED ISSUE #35565: collapse does not work with target IDs with '/'
### Prerequisites

- [X] I have [searched](https://github.com/twbs/bootstrap/issues?utf8=%E2%9C%93&q=is%3Aissue) for duplicate or closed issues
- [X] I have [validated](https://html5.validator.nu/) any HTML to avoid common problems
- [X] I have read the [contributing guidelines](https://github.com/twbs/bootstrap/blob/main/.github/CONTRIBUTING.md)

### Describe the issue

When using collapse with IDs containing '/', the collapse does not work and crashes with the following error:

`Uncaught DOMException: Document.querySelector: '#my/id' is not a valid selector` (tested with FFx, Chrome and Safari)

This is because Document.querySelector expects the query to be properly escaped.

I think this might be fixed at https://github.com/twbs/bootstrap/blob/886b940796b3595a03b44230ca8b78197c5ee1c5/js/src/util/index.js#L55

=> by using something like
```js
if (selector.startsWith("#")) { selector = "#" + CSS.escape(selector.substring(1)) }
return selector
```

=> which would escape properly the selector in a suitable way for document.querySelector() to work properly

### Reduced test cases

Minimal test case (can be pasted in https://codepen.io/team/bootstrap/pen/qBamdLj):

```html
<!-- Follow these instructions, please! -->
<div class="container py-5">
  <h1 class="h2">Bootstrap v5.0.2 Bug Report</h1>
  <p>REPORT: IDs containing slashes do crash. Such IDs need to be escaped before going thu document.querySelector()</p>
  <p>
  <a class="btn btn-primary" data-bs-toggle="collapse" href="#my/id" role="button" aria-expanded="false" aria-controls="my/id">
    Link with href with my/id
  </a>
  <button class="btn btn-primary" type="button" data-bs-toggle="collapse" data-bs-target="#my/id" aria-expanded="false" aria-controls="my/id">
    Button with data-bs-target with my/id
  </button>
</p>
<div class="collapse" id="my/id">
  <div class="card card-body">
    Some placeholder content for the collapse component. This panel is hidden by default but revealed when the user activates the relevant trigger.
  </div>
</div>
</div>
```

### What operating system(s) are you seeing the problem on?

macOS

### What browser(s) are you seeing the problem on?

_No response_

### What version of Bootstrap are you using?

v5.0.2

### LINKED ISSUE #36939: Scrollspy doesn't work with CJK, Greek language targets
### Prerequisites

- [X] I have [searched](https://github.com/twbs/bootstrap/issues?utf8=%E2%9C%93&q=is%3Aissue) for duplicate or closed issues
- [X] I have [validated](https://html5.validator.nu/) any HTML to avoid common problems
- [X] I have read the [contributing guidelines](https://github.com/twbs/bootstrap/blob/main/.github/CONTRIBUTING.md)

### Describe the issue

Related issue twbs/bootstrap#10205(v2/v3)

```text
Uncaught DOMException: Failed to execute 'querySelector' on 'Element': '#%E8%83%8C%E6%99%AF%E5%9B%BE%E7%89%87' is not a valid selector.
    at Object.findOne (https://cdn.jsdelivr.net/npm/bootstrap@5.2.0/dist/js/bootstrap.bundle.min.js:6:9928)
    at mn._initializeTargetsAndObservables (https://cdn.jsdelivr.net/npm/bootstrap@5.2.0/dist/js/bootstrap.bundle.min.js:6:72435)
    at mn.refresh (https://cdn.jsdelivr.net/npm/bootstrap@5.2.0/dist/js/bootstrap.bundle.min.js:6:70756)
    at new mn (https://cdn.jsdelivr.net/npm/bootstrap@5.2.0/dist/js/bootstrap.bundle.min.js:6:70629)
    at mn.getOrCreateInstance (https://cdn.jsdelivr.net/npm/bootstrap@5.2.0/dist/js/bootstrap.bundle.min.js:6:8197)
    at https://cdn.jsdelivr.net/npm/bootstrap@5.2.0/dist/js/bootstrap.bundle.min.js:6:73489
    at i (https://cdn.jsdelivr.net/npm/bootstrap@5.2.0/dist/js/bootstrap.bundle.min.js:6:4253)
```

### Reduced test cases

https://codepen.io/razonyang/pen/rNdZBjg

### What operating system(s) are you seeing the problem on?

Windows

### What browser(s) are you seeing the problem on?

Chrome

### What version of Bootstrap are you using?

v5.2

### PR DESCRIPTION
**Pull request by pierresouchay**
IDs need to be escaped before document.querySelector() is called

Will fix #35565

Closes #34412  
closes #36939
Fixes #34381
Fixes #31099

### CODE REVIEW

**Comment by GeoSot** [comment]
Nice job @pierresouchay . I 've done some testing it and seems it brings another 'feature', enabling use ids starting with number.

Please add some tests to support your MR 😃

**Comment by pierresouchay** [comment]
@GeoSot can you approve the worklow, so I can test it does not break anything?
I'll add the test, for now I just tested in blind mode

**Comment by GeoSot** [comment]
You can try `npm run js-debug` and `npm run js-lint` before pushing, too 😉

**Comment by pierresouchay** [comment]
@GeoSot Thanks, I'll try to add the test this evening. Do you have suggestion on what to test? I think I'll simply change a collapse test to target 0/my/id instead of test2 id...

**Comment by pierresouchay** [comment]
@GeoSot all done, I applied linter, added test with problematic ID and rebased

**Comment by pierresouchay** [comment]
WIll also probably fix https://github.com/twbs/bootstrap/issues/31099

**Comment by GeoSot** [comment]
You probably need to add a test on `index.spec.js` => `getSelector`  and to handle the same thing in the `getElement` function (plus proper test) 😃

**Comment by pierresouchay** [comment]
> You probably need to add a test on `index.spec.js` => `getSelector` and to handle the same thing in the `getElement` function (plus proper test) 😃

Done in last commit

**Comment by pierresouchay** [comment]
@GeoSot or @XhmikosR can you launch the tests, I think the PR is ready.

Regards

**Comment by pierresouchay** [comment]
@GeoSot Current support is not bad https://caniuse.com/mdn-api_css_escape 

Capture d’écran 2021-12-20 à 09 48 57

Does it fit the currently supported browsers for Bootstrap?

**Comment by pierresouchay** [comment]
@GeoSot I could do another minor change (if you want me to):

```js
  if (selector !== null && selector.startsWith('#')) {
    // document.querySelector needs escaping to handle IDs (html5+) containing for instance /
    selector = '#' + CSS.escape(selector.slice(1))
  }
```

to be changed to:

```js
  if (selector !== null && selector.startsWith('#') && window.CSS && window.CSS.escape) {
    // document.querySelector needs escaping to handle IDs (html5+) containing for instance /
    selector = '#' + CSS.escape(selector.slice(1))
  }
```

=> so It would work as it does now for very old browsers not supporting CSS.escape (so, would still break on non HTML4 IDs, but would work for HTML4 IDs)

**Comment by pierresouchay** [comment]
Hello @XhmikosR,

I added the tests (as seen in the labels of the PR), do you need more?

(BTW: I need someone to approve the tests in order to run)

**Comment by pierresouchay** [comment]
IMHO, this change fixes bugs, not all of them, ans is straightforward (while we might argue on my comment about detecting the CSS.escape() function)
The escaping of other places with non-leading # is an issue I agree, but it probably far more complex and requires more refactoring. Unfortunately, this would requires some clever parsing because the # is not supported as it is with CSS escape as it requires more context to know if we are talking about ids of "regular" selectors (I mean html 4 compliant ones).
Still I think the / is more and more an issue as it makes sense to have IDs compliant with hash based-routing/history in single web pages (that's why I found the issue in the first place)

**Comment by GeoSot** [comment]
Personally, I am ok with this check 
` (selector && selector.startsWith('#') && window.CSS && window.CSS.escape)`


@XhmikosR , please tell us you opinion. shall we go with the MVP or not?

**Comment by pierresouchay** [comment]
@GeoSot & @XhmikosR Done, updated the test to detected CSS.escape() in latest patch.

Merry Christmas to all of you!

**Comment by pierresouchay** [comment]
@XhmikosR & @GeoSot hello,
Hope you are all doing well. Did you decide something about this PR?

Regards

File: js/tests/unit/tab.spec.js [hunk 1]
```diff
@@ -133,6 +133,43 @@ describe('Tab', () => {
       tab.show()
     })
 
+    it('should work with tab id being an int (fix #31099)', done => {
```
Comment (GeoSot) [inline js/tests/unit/tab.spec.js:136]:
```suggestion
    it('should work with tab id being an int', done => {
```

**Comment by GeoSot** [comment]
I 'll approve it as it solves the 3 issues, keeping in mind that it is not an end up solution (the other 7  occurrences have to be be changed in the future)

**Review by GeoSot — APPROVED** [review]

**Comment by pierresouchay** [comment]
Hello @XhmikosR & @GeoSot

What should I do to get this MR being merged (I know, I have to solve conflicts first)?

Do you want to rewrite all places where a selector is used?

Regards

**Comment by mdo** [comment]
@GeoSot @XhmikosR Is this PR good to go?

**Comment by pierresouchay** [comment]
Ping?

**Comment by XhmikosR** [comment]
Sorry for the late reply, I've been pretty busy with real life.

I'll try to have a look in the next days.

**Comment by pierresouchay** [comment]
Hello @GeoSot & @XhmikosR do you want to consider this? Should I try updating it?

**Comment by GeoSot** [comment]
Hello, we had discussed a bit. 
I can recall that the main thoughts were that :
* we have multiple places we are using the `querySelector` (so would be inconsistent in some cases) 
*  that in case we support this, we have to be ready for new problems in case of selectors like `#2foo .bar` or `#2foo #2bar`

So I really don't have a valid answer right now, except that I am already working on the background to make the decision and the change easier in future #36027

**Comment by larseggert** [comment]
I think we get hit by this, and a fix would be welcome.
```
Uncaught DOMException: Element.querySelector: '#section-1.1' is not a valid selector
```

**Comment by pierresouchay** [comment]
> Hello, we had discussed a bit. I can recall that the main thoughts were that :
> 
>     * we have multiple places we are using the `querySelector` (so would be inconsistent in some cases)
> 
>     * that in case we support this, we have to be ready for new problems in case of selectors like `#2foo .bar` or `#2foo #2bar`
> 
> 
> So I really don't have a valid answer right now, except that I am already working on the background to make the decision and the change easier in future #36027

Hello @GeoSot ,

Ok, I had another proposal, to use the following:

```js
if (selector !== null && window.CSS && window.CSS.escape) {
  selector = selector.replaceAll(/#([^\s#]+)/g, function(full,id){return "#" + CSS.escape(id)})
}
```

A few examples of results:

```
"#section-1.1 .bar".replaceAll(/#([^\s#]+)/g, function(_full,id){return "#" + CSS.escape(id)})
"#section-1\\.1 .bar" 
```

```
"#foo .bar".replaceAll(/#([^\s#]+)/g, function(_full,id){return "#" + CSS.escape(id)})
=> "#foo .bar"
```

It sounds to work well... but as you mentioned, for some weird reason, it would not work for IDs starting with numbers.

So, for some weird reason, CSS.escape('2f') => "32 f"

So, in your example:

```
"#2foo .bar".replaceAll(/#([^\s#]+)/g, function(_full,id){return "#" + CSS.escape(id)})
=> "#\\32 foo .bar" => BAD
```

From my testing, this is only an issue with identifiers starting with numbers. I don't have a clear understanding of why.

What I thus propose is to basically ignore IDs starting with a number and thus to use:

```js
if (selector !== null && window.CSS && window.CSS.escape) {
  selector = selector.replaceAll(/#([^0-9][^\s#]+)/g, function(full,id){return "#" + CSS.escape(id)})
}
```

Which gives some more correct results than the current implementation:
```
"#2foo .bar > #this-is-my-solution-1.0".replaceAll(/#([^0-9][^\s#]+)/g, function(_full,id){return "#" + CSS.escape(id)})
=> "#2foo .bar > #this-is-my-solution-1\\.0" 
```

The only downside is that it does not work completely well with IDs starting with numbers, but for everything else, it works.

What do you think?

**Comment by pierresouchay** [comment]
@GeoSot huh... according to https://mothereff.in/css-escapes and my test, the CSS.escape is actually working!

So, I am gonna submit a new patch

**Comment by pierresouchay** [comment]
@GeoSot @XhmikosR Ok, I am now happy with this new version, it now works by escaping all IDs, whatever their position in the selector

Works with complex selectors such as ".foo 2bar", added a UTest to ensure it works well: https://github.com/twbs/bootstrap/pull/35566/files#diff-232a393b53d525c6749638629be8a4288ec1929b51c8144fe92a41c25c885bc3R132

**Comment by XhmikosR** [comment]
We need to double check if our currently supported browsers have this feature.

Apart from that:

1. as long as the change covers all cases
2. everything is covered by tests

we could try landing it.

**Comment by pierresouchay** [comment]
@XhmikosR I documented what is supported: https://caniuse.com/mdn-api_css_escape

Since a feature tested, older systems will behave as they do now

**Comment by pierresouchay** [comment]
@GeoSot @XhmikosR Could someone validate the workflow so all tests do converge? (I did rebase to be up-to-date with current main branch...)

**Comment by pierresouchay** [comment]
Hello @GeoSot and @XhmikosR ,

- This PR will be soon 11 months old.
- It fixes lots of issues.
- It is simple enough.
- it has tests that confirm the fixes
- It degrades if feature is not supported.
- I rebased several times to have something that could be merged easily. 

What is missing?

**Review by GeoSot — CHANGES_REQUESTED** [review]
I am just ok with the index.js test. Let's keep it simple

> What is missing?


It doesn't miss something, and thank you a lot for your patience. The only thing is that is a difficult decision as it may impact many different cases, we haven't thought of. 

For me, it seems ok to go. Worst case scenario, we will revert it

**Comment by GeoSot** [comment]
@pierresouchay I've rebased the branch, according to the last changes, but seems I did something wrong with the tab test. Can you please take a look?

File: .bundlewatch.config.json [hunk 2]
```diff
@@ -38,7 +38,7 @@
     },
     {
       "path": "./dist/js/bootstrap.bundle.min.js",
-      "maxSize": "22.75 kB"
+      "maxSize": "22.80 kB"
```
Comment (julien-deramond) [inline .bundlewatch.config.json:41]:
```suggestion
      "maxSize": "23.0 kB"
```

**Review by GeoSot — APPROVED** [review]

**Comment by GeoSot** [comment]
No, we are done ;) 

Thank you a lot for your patience and your time

**Comment by XhmikosR** [comment]
@pierresouchay you should make a separate issue with all the info there.

**Comment by larseggert** [comment]
@XhmikosR adding this to a test illustrates the bug:
```patch
diff --git a/js/tests/unit/scrollspy.spec.js b/js/tests/unit/scrollspy.spec.js
index c7951e6..704b52a 100644
--- a/js/tests/unit/scrollspy.spec.js
+++ b/js/tests/unit/scrollspy.spec.js
@@ -341,11 +341,13 @@ describe('ScrollSpy', () => {
           '  <ul class="nav">',
           '    <li class="nav-item"><a class="nav-link" id="a-1" href="#div-1">div 1</a></li>',
           '    <li class="nav-item"><a class="nav-link" id="a-2" href="#div-2">div 2</a></li>',
+          '    <li class="nav-item"><a class="nav-link" id="a-2.1" href="#div-2.1">div 2</a></li>',
           '  </ul>',
           '</nav>',
           '<div class="content" style="overflow: auto; height: 50px">',
           '  <div id="div-1" style="height: 100px; padding: 0; margin: 0">div 1</div>',
           '  <div id="div-2" style="height: 200px; padding: 0; margin: 0">div 2</div>',
+          '  <div id="div-2.1" style="height: 200px; padding: 0; margin: 0">div 2.1</div>',
           '</div>'
         ].join('')
```

**Comment by julien-deramond** [comment]
@larseggert Could you please create a separate issue with this information and the problem you have? Otherwise, it is very difficult for us to track since this PR is already merged.

**Comment by larseggert** [comment]
See #37858

**Pull request merged**